### PR TITLE
feat: support automatic IAM authentication

### PR DESCRIPTION
**Pull request by shubha-rajan**
depends on #84 
- Add AuthType parameter to connection settings
- Append access token to ephemeral cert request when IAM auth is enabled.

### CODE REVIEW

File: test/connector.ts [hunk 1]
```diff
@@ -58,7 +58,8 @@ t.test('Connector', async t => {
 
   const connector = new Connector();
   const opts = await connector.getOptions({
-    type: 'PUBLIC',
+    ipType: 'PUBLIC',
```
Comment (enocom) [inline test/connector.ts:61]: Let's split out the breaking changes into a separate PR.

File: test/connector.ts [hunk 2]
```diff
@@ -72,11 +73,11 @@ t.test('Connector invalid type error', async t => {
   t.rejects(
     connector.getOptions({
       ipType: 'foo' as IpAdressesTypes,
-      authType: AuthTypes.PASSWORD,
+      authType: 'PASSWORD' as AuthTypes,
```
Comment (enocom) [inline test/connector.ts:76]: Let's avoid `as`. See https://google.github.io/styleguide/tsguide.html#type-and-non-nullability-assertions.

File: src/connector.ts [hunk 3]
```diff
@@ -52,29 +55,40 @@ export declare interface DriverOptions {
...
-    type: IpAdressesTypes;
+    ipType: IpAdressesTypes;
+    authType: AuthTypes;
     instanceConnectionName: string;
     sqlAdminFetcher: SQLAdminFetcher;
   }): Promise<void> {
     // in case an instance to that connection name has already
     // been setup there's no need to set it up again
-    if (this.has(instanceConnectionName)) {
+    if (this.has(this.getKey({instanceConnectionName, ipType, authType}))) {
```
Comment (enocom) [inline src/connector.ts:70]: This should be a separate PR as it's a departure from other connectors.

Comment (ruyadorno) [reply src/connector.ts:70]: curious, how does the other connectors handle possible concurrent ip/auth types within a single instance?

Comment (enocom) [reply src/connector.ts:70]: They don't handle it. The direction here is correct, but I'd like to land this change in Go first and with some thought into the implications of this change.

Comment (jackwotherspoon) [reply src/connector.ts:70]: python connector handles this by throwing an error if `enable_iam_auth` argument has changed since last `connect` call. https://github.com/GoogleCloudPlatform/cloud-sql-python-connector/blob/main/google/cloud/sql/connector/connector.py#L188-L197

Comment (enocom) [reply src/connector.ts:70]: This is a good example of where we need to bring all the connectors in line feature-wise. But again not in this PR. ;-)

File: src/connector.ts [hunk 4]
```diff
@@ -52,29 +55,40 @@ export declare interface DriverOptions {
...
       return;
     }
     const connectionInstance = await CloudSQLInstance.getCloudSQLInstance({
-      connectionType: type,
+      connectionType: ipType,
+      authType: authType,
```
Comment (ruyadorno) [inline src/connector.ts:75]: when the key name and the assigning value var name match, you can simplify the declaration by just declaring that name once:

```suggestion
      authType,
```

File: src/sqladmin-fetcher.ts [hunk 5]
```diff
@@ -99,14 +99,34 @@ export class SQLAdminFetcher {
...
     {projectId, instanceId}: InstanceConnectionInfo,
-    publicKey: string
+    publicKey: string,
+    auth?: GoogleAuth
   ): Promise<SslCert> {
+
+    type RequestBody = {
+      public_key: string,
+      access_token?: string
+    }
```
Comment (ruyadorno) [inline src/sqladmin-fetcher.ts:109]: we should favour `interface` declarations over `type`, ref: https://google.github.io/styleguide/tsguide.html#interfaces-vs-type-aliases

I would also prefer to yank this to ~L28 so that it's declared right after the other interface declaration there 😊

File: test/connector.ts (hunk 2 above)
Comment (ruyadorno) [reply test/connector.ts:76]: in tests it's not so bad if needed (to use `as`) but yeah, in this specific case here though, it seems to me that it's easier to import the types and use them directly

File: test/sqladmin-fetcher.ts [hunk 6]
```diff
@@ -64,7 +64,9 @@ const mockRequest = (
       ...overrides,
     });
 };
-
+t.beforeEach(() => {
+  nock.cleanAll();
+})
```
Comment (ruyadorno) [inline test/sqladmin-fetcher.ts:69]: This can be removed 😊

File: src/connector.ts (hunk 3 above)
Comment (ruyadorno) [reply src/connector.ts:70]: I have the impression the follow up PRs are going to be quite small once we remove this and split the proposed changes appart 😊

File: src/cloud-sql-instance.ts [hunk 7]
```diff
@@ -49,23 +51,33 @@ export class CloudSQLInstance {
...
     this.connectionType = connectionType;
+    this.authType = authType;
     this.instanceInfo = parseInstanceConnectionName(instanceConnectionName);
     this.sqlAdminFetcher = sqlAdminFetcher;
+
+    if (authType === AuthTypes.IAM) {
+      this.auth = new GoogleAuth({
+        scopes: ["https://www.googleapis.com/auth/sqlservice.login"],
+      });
+    }
```
Comment (ruyadorno) [inline src/cloud-sql-instance.ts:80]: Do we really need an independent `GoogleAuth` instance? Looking at where it's used in `src/sqladmin-fetcher.ts` it seems to me that this new instance is only being used to call `auth.getAccessToken()`.

If that's the case it seems much simpler to [keep a reference to the `GoogleAuth` instance declared in the constructor of the `SQLAdminFetcher` class](https://github.com/GoogleCloudPlatform/cloud-sql-nodejs-connector/blob/93168049995520bc290368b0530fd768d052db38/src/sqladmin-fetcher.ts#L33) and simply call that method in that same instance on `getEphemeralCertificate`. Which should avoid all changes to `src/cloud-sql-instance.ts`.

File: test/sqladmin-fetcher.ts
Comment (ruyadorno) [inline test/sqladmin-fetcher.ts:1]: I'd like to see a test asserting that the `access_token` is properly set in the request and also another one showing the `ENOACCESSTOKEN` error happening

**Review by ruyadorno — COMMENTED** [review]
It looks like there are a few small lint errors here and there, these should be mostly taken care of by running `npm run fix`, if there are more you can always check by running `npm run lint` 😊 

The direction here LGTM assuming that I understand correctly how the IAM auth should work, feel free to correct my comments if some of my assumptions are wrong. 

Thanks @shubha-rajan for the contribution!

File: src/connector.ts (hunk 3 above)
Comment (enocom) [reply src/connector.ts:70]: Little follow up -- I think we want to be following Python here. But still probably better as a separate PR after a one-pager looking at all the connectors.

File: test/sqladmin-fetcher.ts
Comment (shubha-rajan) [reply test/sqladmin-fetcher.ts:1]: Added these tests!

File: src/cloud-sql-instance.ts (hunk 7 above)
Comment (shubha-rajan) [reply src/cloud-sql-instance.ts:80]: So the reason we're creating a new Auth instance here is that this instance has more restricted scopes (only `sqlservice.login`, while the one in `src/sqladmin-fetcher.ts` has `sqlservice.admin`

File: src/connector.ts (hunk 4 above)
Comment (shubha-rajan) [reply src/connector.ts]: fixed!

File: src/sqladmin-fetcher.ts (hunk 5 above)
Comment (shubha-rajan) [reply src/sqladmin-fetcher.ts]: done!

File: src/auth-types.ts
Comment (ruyadorno) [inline src/auth-types.ts:1]: this enum should be imported/exported in `src/index.ts`, similar to `IpAddressTypes`

File: src/cloud-sql-instance.ts (hunk 7 above)
Comment (ruyadorno) [reply src/cloud-sql-instance.ts:80]: I see! that makes sense, I still wonder though if it wouldn't be simpler than to have a single extra `GoogleAuth` instance living in `src/sqladmin-fetcher.ts` (something like `this.iamAuth`) instead of adding a new `auth` reference for each cloud sql instance

Comment (shubha-rajan) [reply src/cloud-sql-instance.ts:80]: I think that's a good idea, as long as we'll never run into a case where we'd need to use different auth sources with the same connector. I don't think it will be a problem though.

File: src/sqladmin-fetcher.ts [hunk 8]
```diff
@@ -131,9 +155,17 @@ export class SQLAdminFetcher {
...
+    if (tokenExpiration) {
+      if (tokenExpiration < nearestExpiration) {
+        nearestExpiration = tokenExpiration;
+      }
+    }
+
     return {
       cert,
-      expirationTime,
+      expirationTime: new Date(nearestExpiration).toISOString(),
```
Comment (ruyadorno) [inline src/sqladmin-fetcher.ts:168]: this will be better handled as a sep util function in `src/time.ts`

File: src/cloud-sql-instance.ts (hunk 7 above)
Comment (ruyadorno) [reply src/cloud-sql-instance.ts:80]: cool 👍 yeah, let's move it there then, I'm not really liking the idea of adding `GoogleAuth` into `src/cloud-sql-instance.ts`

Comment (jackwotherspoon) [reply src/cloud-sql-instance.ts:80]: We may want to discuss this before implementing it for feature parity across connectors. As I know atleast for the Python Connector we currently attach a Google Auth credentials object to each instance at the instance level. It might be a good enhancement across the board to move the Google Auth credentials up a level to the Connector level but we should probably discuss it a bit more to make sure there isn't a use case for leaving it per instance.

Comment (shubha-rajan) [reply src/cloud-sql-instance.ts:80]: Java is also at the instance level (because before the refactor there wasn't a connector level). But I agree, let's leave it as-is for now and discuss this more.

Comment (ruyadorno) [reply src/cloud-sql-instance.ts:80]: Having in mind that the go connector is the reference implementation it seems to me like an additional argument to move it out of the instance level.

**Review by ruyadorno — COMMENTED** [review]
Let's also add to the top of the list of benefits in the **README** file:

```patch
+* **IAM Authorization:** uses IAM permissions to control who/what can connect to
+  your Cloud SQL instances
```

File: src/time.ts [hunk 9]
```diff
@@ -33,3 +33,13 @@ export function getRefreshInterval(isoTime: string): number {
...
+
+export function getNearestExpiration(
+  certExp: number,
+  tokenExp: number | undefined
+) {
+  if (tokenExp) {
+    return new Date(Math.min(certExp, tokenExp)).toISOString();
+  }
+  return new Date(certExp).toISOString();
+}
```
Comment (ruyadorno) [inline src/time.ts:45]: it would be nice to add some unit tests for this method, should be fairly straightforward now that it's on `src/time.ts` and might help out shaking off any edge cases

File: test/sqladmin-fetcher.ts [hunk 10]
```diff
@@ -247,7 +248,7 @@ t.test('getEphemeralCertificate', async t => {
     ephemeralCert,
     {
       cert: CLIENT_CERT,
-      expirationTime: 'Jul 22 17:53:09 3022 GMT',
+      expirationTime: '3022-07-22T17:53:09.000Z',
```
Comment (ruyadorno) [inline test/sqladmin-fetcher.ts:251]: interesting, this seems unrelated to the auth type changes, why did it needed to change?

Comment (shubha-rajan) [reply test/sqladmin-fetcher.ts:251]: This is because we're using `toISOString()` before we return the expiration time now.

Comment (ruyadorno) [reply test/sqladmin-fetcher.ts:251]: oh I see, yeah that makes sense!

**Comment by jackwotherspoon** [comment]
@shubha-rajan curious if there is a plan to add system tests for IAM AuthN as part of this PR?

**Comment by shubha-rajan** [comment]
> @shubha-rajan curious if there is a plan to add system tests for IAM AuthN as part of this PR?

Gonna do it in the next PR! This one's already gotten bigger than I expected.

File: src/cloud-sql-instance.ts [hunk 11]
```diff
@@ -13,13 +13,15 @@
 // limitations under the License.
 
 import {IpAddressTypes, selectIpAddress} from './ip-addresses';
+import {GoogleAuth} from 'google-auth-library';
```
Comment (jackwotherspoon) [inline src/cloud-sql-instance.ts:16]: nit: can we get rid of this import now that we moved auth to the connector level?
```suggestion
```

**Review by jackwotherspoon — APPROVED** [review]
LGTM. one small nit about unused import

File: src/connector.ts [hunk 12]
```diff
@@ -53,30 +55,41 @@ export declare interface DriverOptions {
...
+    authType: AuthTypes;
     instanceConnectionName: string;
     sqlAdminFetcher: SQLAdminFetcher;
   }): Promise<void> {
     // in case an instance to that connection name has already
     // been setup there's no need to set it up again
     if (this.has(instanceConnectionName)) {
       const instance = this.get(instanceConnectionName);
-      if (instance.ipType !== ipType) {
+      if (instance.ipType && instance.ipType !== ipType) {
```
Comment (ruyadorno) [inline src/connector.ts:71]:
```suggestion
      if (instance.ipType !== ipType) {
```

I don't see why the extra check is necessary, can we remove it?

File: src/connector.ts [hunk 13]
```diff
@@ -53,30 +55,41 @@ export declare interface DriverOptions {
...
         throw new CloudSQLConnectorError({
           message:
             `getOptions called for instance ${instanceConnectionName} with ipType ${ipType}, ` +
             `but was previously called with ipType ${instance.ipType}. ` +
             'If you require both for your use case, please use a new connector object.',
           code: 'EMISMATCHIPTYPE',
         });
+      } else if (instance.authType && instance.authType !== authType) {
```
Comment (ruyadorno) [inline src/connector.ts:79]:
```suggestion
      } else if (instance.authType !== authType) {
```

similar to the previous comment

File: src/connector.ts (hunk 12 above)
Comment (ruyadorno) [reply src/connector.ts:71]: nvm! I just run the tests and I see now that we need that extra check to see if `ipType` is defined 😬
